aws_crt_common_ws/aws-crt-common/src/lib.rs

```diff
@@ -0,0 +1,596 @@
+use std::future::Future;
+use std::os::raw::{c_char, c_void};
+use std::pin::Pin;
+use std::ptr::null_mut;
+use std::str::Utf8Error;
+use std::sync::{Arc, Mutex, MutexGuard};
+use std::task::{Context, Poll};
+
+/// Trait for deep copying (cloning) an object. The operation can fail so it returns
+/// a result with an error code upon failure.
+pub trait TryClone<T> {
+    fn try_clone(&self) -> Result<T, i32>;
+}
+
+/// Trait that gives a c-style non-const pointer to an underlying resource.
+/// Ideally you use this when calling an ffi function that needs a pointer
+/// to your wrapped Rust object.
+pub trait MutableNativeResource<T> {
+    fn get_native_resource_mut(&mut self) -> *mut T;
+}
+
+/// Trait that gives a c-style const pointer to an underlying resource.
+/// Ideally you use this when calling an ffi function that needs a pointer
+/// to your wrapped Rust object.
+pub trait NativeResource<T> {
+    fn get_native_resource(&self) -> *const T;
+}
+
+pub struct FutureSharedState<T>
+where
+    T: Clone,
+{
+    data: Option<T>,
+    waker: Option<Waker>,
+}
+
+/// Simple future implementation for driving async futures
+/// from C-style callbacks. For this scenario, a pointer to FutureSharedState is used as your user_data pointer.
+///
+/// Upon creation call, CallbackFuture::to_callback_ptr() to get the pointer to pass to C functions.
+/// When your callback completes, fill-in an instance of type T, and call FutureSharedState::resolve()
+/// with the user_data pointer from your c callback.
+///
+/// Assuming you do these things, you've just converted C-style async code into co-routine ready
+/// Rust futures.
+///
+#[derive(Clone)]
+pub struct CallbackFuture<T>
+where
+    T: Clone,
+{
+    data: Arc<Mutex<FutureSharedState<T>>>,
+}
+
+impl<T> FutureSharedState<T>
+where
+    T: Clone,
+{
+    ///
+    /// Completes the future for callback_data with result as the future's output.
+    ///
+    pub fn resolve(callback_data: *const c_void, result: T) {
+        // sound assuming a valid instance of this class.
+        unsafe {
+            let shared_state_ptr = Arc::from_raw(callback_data as *mut Mutex<FutureSharedState<T>>);
+
+            let mut res_option = shared_state_ptr.lock().unwrap();
+            res_option.data = Some(result);
+
+            if let Some(waker) = res_option.waker.take() {
+                waker.wake();
+            }
+        }
+    }
+}
+
+impl<T> Future for CallbackFuture<T>
+where
+    T: Clone,
+{
+    type Output = T;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut option_val: MutexGuard<FutureSharedState<T>> = self.data.lock().unwrap();
+        if option_val.data.is_none() {
+            option_val.waker = Some(cx.waker().clone());
+            return Poll::Pending;
+        }
+
+        return Poll::Ready(option_val.data.as_ref().expect("We're in a ready state implying the future has completed but the underlying value hasn't been set!").clone());
```

Comment:
use `option_val.data.take()` and remove the `Clone` bound

Reply:
https://github.com/awslabs/aws-crt-rs/compare/russell/aws-string-safety?expand=1#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759L101-L112